From 9a49f61f5d60db0a0c1839a97f20baa0ca6f2f0c Mon Sep 17 00:00:00 2001
From: RJ Sampson <rj.sampson@chainguard.dev>
Date: Tue, 3 Mar 2026 05:20:58 -0700
Subject: [PATCH 1/7] fix(cli/keystore): Set BCFIPS-compliant password by default

This changes default behavior of the keystore CLI so that a BCFIPS-compliant
password, "chageitchangeit", is set on keystore creation. On initial create,
we also create the file used by ECK operator to flag that a keystore has
been created so that it does not attempt to overwrite it (if a password
is set on the keystore, ECK operator attempts to create a new keystore on
init).

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>
---
 .../cli/keystore/BaseKeyStoreCommand.java     |  5 ++--
 .../cli/keystore/CreateKeyStoreCommand.java   | 11 ++++++-
 .../server/cli/KeyStoreLoader.java            | 15 +++++++---
 .../elasticsearch/server/cli/ServerCli.java   |  3 +-
 .../common/cli/KeyStoreAwareCommand.java      |  3 +-
 .../common/settings/KeyStoreConfig.java       | 30 +++++++++++++++++++
 .../common/settings/KeyStoreWrapper.java      |  4 +--
 7 files changed, 60 insertions(+), 11 deletions(-)
 create mode 100644 server/src/main/java/org/elasticsearch/common/settings/KeyStoreConfig.java

diff --git a/distribution/tools/keystore-cli/src/main/java/org/elasticsearch/cli/keystore/BaseKeyStoreCommand.java b/distribution/tools/keystore-cli/src/main/java/org/elasticsearch/cli/keystore/BaseKeyStoreCommand.java
index a8a75ac23c9..0d5dbc33f55 100644
--- a/distribution/tools/keystore-cli/src/main/java/org/elasticsearch/cli/keystore/BaseKeyStoreCommand.java
+++ b/distribution/tools/keystore-cli/src/main/java/org/elasticsearch/cli/keystore/BaseKeyStoreCommand.java
@@ -17,6 +17,7 @@ import org.elasticsearch.cli.ProcessInfo;
 import org.elasticsearch.cli.Terminal;
 import org.elasticsearch.cli.UserException;
 import org.elasticsearch.common.cli.KeyStoreAwareCommand;
+import org.elasticsearch.common.settings.KeyStoreConfig;
 import org.elasticsearch.common.settings.KeyStoreWrapper;
 import org.elasticsearch.common.settings.SecureString;
 import org.elasticsearch.env.Environment;
@@ -55,11 +56,11 @@ public abstract class BaseKeyStoreCommand extends KeyStoreAwareCommand {
                         return;
                     }
                 }
-                keyStorePassword = new SecureString(new char[0]);
+                keyStorePassword = new SecureString(KeyStoreConfig.getPasswordChars());
                 keyStore = KeyStoreWrapper.create();
                 keyStore.save(configFile, keyStorePassword.getChars());
             } else {
-                keyStorePassword = keyStore.hasPassword() ? readPassword(terminal, false) : new SecureString(new char[0]);
+                keyStorePassword = keyStore.hasPassword() ? readPassword(terminal, false) : new SecureString(KeyStoreConfig.getPasswordChars());
                 keyStore.decrypt(keyStorePassword.getChars());
             }
             executeCommand(terminal, options, env);
diff --git a/distribution/tools/keystore-cli/src/main/java/org/elasticsearch/cli/keystore/CreateKeyStoreCommand.java b/distribution/tools/keystore-cli/src/main/java/org/elasticsearch/cli/keystore/CreateKeyStoreCommand.java
index ef561b08d9a..bed5f4f21dc 100644
--- a/distribution/tools/keystore-cli/src/main/java/org/elasticsearch/cli/keystore/CreateKeyStoreCommand.java
+++ b/distribution/tools/keystore-cli/src/main/java/org/elasticsearch/cli/keystore/CreateKeyStoreCommand.java
@@ -17,6 +17,7 @@ import org.elasticsearch.cli.ProcessInfo;
 import org.elasticsearch.cli.Terminal;
 import org.elasticsearch.cli.UserException;
 import org.elasticsearch.common.cli.KeyStoreAwareCommand;
+import org.elasticsearch.common.settings.KeyStoreConfig;
 import org.elasticsearch.common.settings.KeyStoreWrapper;
 import org.elasticsearch.common.settings.SecureString;
 import org.elasticsearch.env.Environment;
@@ -37,15 +38,23 @@ class CreateKeyStoreCommand extends KeyStoreAwareCommand {
         this.passwordOption = parser.acceptsAll(Arrays.asList("p", "password"), "Prompt for password to encrypt the keystore");
     }
 
+    // This file must exist after init keystore creation or ECK operator will attempt to recreate the keystore
+    private static final String ECK_INIT_KEYSTORE_FLAG = "elastic-internal-init-keystore.ok";
+
     @Override
     public void execute(Terminal terminal, OptionSet options, Environment env, ProcessInfo processInfo) throws Exception {
-        try (SecureString password = options.has(passwordOption) ? readPassword(terminal, true) : new SecureString(new char[0])) {
+        final char[] passwordChars = KeyStoreConfig.getPasswordChars();
+        try (SecureString password = options.has(passwordOption) ? readPassword(terminal, true) : new SecureString(passwordChars)) {
             Path keystoreFile = KeyStoreWrapper.keystorePath(env.configDir());
             if (Files.exists(keystoreFile)) {
                 if (terminal.promptYesNo("An elasticsearch keystore already exists. Overwrite?", false) == false) {
                     terminal.println("Exiting without creating keystore.");
                     return;
                 }
+            } else {
+                Path flagFile = env.configDir().resolve(ECK_INIT_KEYSTORE_FLAG);
+                Files.createDirectories(flagFile.getParent());
+                Files.writeString(flagFile, "");
             }
             KeyStoreWrapper keystore = KeyStoreWrapper.create();
             keystore.save(env.configDir(), password.getChars());
diff --git a/distribution/tools/server-cli/src/main/java/org/elasticsearch/server/cli/KeyStoreLoader.java b/distribution/tools/server-cli/src/main/java/org/elasticsearch/server/cli/KeyStoreLoader.java
index 2ae58040437..0b087bbe5ba 100644
--- a/distribution/tools/server-cli/src/main/java/org/elasticsearch/server/cli/KeyStoreLoader.java
+++ b/distribution/tools/server-cli/src/main/java/org/elasticsearch/server/cli/KeyStoreLoader.java
@@ -10,6 +10,7 @@
 package org.elasticsearch.server.cli;
 
 import org.elasticsearch.cli.Terminal;
+import org.elasticsearch.common.settings.KeyStoreConfig;
 import org.elasticsearch.common.settings.KeyStoreWrapper;
 import org.elasticsearch.common.settings.SecureSettings;
 import org.elasticsearch.common.settings.SecureString;
@@ -25,10 +26,16 @@ public class KeyStoreLoader implements SecureSettingsLoader {
     public LoadedSecrets load(Environment environment, Terminal terminal) throws Exception {
         // See if we have a keystore already present
         KeyStoreWrapper secureSettings = KeyStoreWrapper.load(environment.configDir());
-        // If there's no keystore or the keystore has no password, set an empty password
-        var password = (secureSettings == null || secureSettings.hasPassword() == false)
-            ? new SecureString(new char[0])
-            : new SecureString(terminal.readSecret(KeyStoreWrapper.PROMPT));
+        SecureString password;
+        if (secureSettings != null && secureSettings.hasPassword()) {
+            password = new SecureString(terminal.readSecret(KeyStoreWrapper.PROMPT));
+            // Set default KeyStore password if blank
+            if (password.length() == 0 || password.toString().isBlank()) {
+                password = new SecureString(KeyStoreConfig.getPasswordChars());
+            }
+        } else {
+            password = new SecureString(KeyStoreConfig.getPasswordChars());
+        }
 
         return new LoadedSecrets(secureSettings, Optional.of(password));
     }
diff --git a/distribution/tools/server-cli/src/main/java/org/elasticsearch/server/cli/ServerCli.java b/distribution/tools/server-cli/src/main/java/org/elasticsearch/server/cli/ServerCli.java
index 358358c8140..895cbc45a22 100644
--- a/distribution/tools/server-cli/src/main/java/org/elasticsearch/server/cli/ServerCli.java
+++ b/distribution/tools/server-cli/src/main/java/org/elasticsearch/server/cli/ServerCli.java
@@ -23,6 +23,7 @@ import org.elasticsearch.cli.ProcessInfo;
 import org.elasticsearch.cli.Terminal;
 import org.elasticsearch.cli.UserException;
 import org.elasticsearch.common.cli.EnvironmentAwareCommand;
+import org.elasticsearch.common.settings.KeyStoreConfig;
 import org.elasticsearch.common.settings.SecureSettings;
 import org.elasticsearch.common.settings.SecureString;
 import org.elasticsearch.env.Environment;
@@ -85,7 +86,7 @@ class ServerCli extends EnvironmentAwareCommand {
 
         try (
             var loadedSecrets = secureSettingsLoader.load(env, terminal);
-            var password = (loadedSecrets.password().isPresent()) ? loadedSecrets.password().get() : new SecureString(new char[0]);
+            var password = (loadedSecrets.password().isPresent()) ? loadedSecrets.password().get() : new SecureString(KeyStoreConfig.getPasswordChars());
         ) {
             SecureSettings secrets = loadedSecrets.secrets();
             if (secureSettingsLoader.supportsSecurityAutoConfiguration()) {
diff --git a/server/src/main/java/org/elasticsearch/common/cli/KeyStoreAwareCommand.java b/server/src/main/java/org/elasticsearch/common/cli/KeyStoreAwareCommand.java
index 7506ce18daa..8be9f526946 100644
--- a/server/src/main/java/org/elasticsearch/common/cli/KeyStoreAwareCommand.java
+++ b/server/src/main/java/org/elasticsearch/common/cli/KeyStoreAwareCommand.java
@@ -15,6 +15,7 @@ import org.elasticsearch.cli.ExitCodes;
 import org.elasticsearch.cli.ProcessInfo;
 import org.elasticsearch.cli.Terminal;
 import org.elasticsearch.cli.UserException;
+import org.elasticsearch.common.settings.KeyStoreConfig;
 import org.elasticsearch.common.settings.KeyStoreWrapper;
 import org.elasticsearch.common.settings.SecureString;
 import org.elasticsearch.env.Environment;
@@ -63,7 +64,7 @@ public abstract class KeyStoreAwareCommand extends EnvironmentAwareCommand {
      * Decrypt the {@code keyStore}, prompting the user to enter the password in the {@link Terminal} if it is password protected
      */
     protected static void decryptKeyStore(KeyStoreWrapper keyStore, Terminal terminal) throws UserException, IOException {
-        try (SecureString keystorePassword = keyStore.hasPassword() ? readPassword(terminal, false) : new SecureString(new char[0])) {
+        try (SecureString keystorePassword = keyStore.hasPassword() ? readPassword(terminal, false) : new SecureString(KeyStoreConfig.getPasswordChars())) {
             keyStore.decrypt(keystorePassword.getChars());
         } catch (SecurityException | GeneralSecurityException e) {
             throw new UserException(ExitCodes.DATA_ERROR, e.getMessage());
diff --git a/server/src/main/java/org/elasticsearch/common/settings/KeyStoreConfig.java b/server/src/main/java/org/elasticsearch/common/settings/KeyStoreConfig.java
new file mode 100644
index 00000000000..de0efed5f7b
--- /dev/null
+++ b/server/src/main/java/org/elasticsearch/common/settings/KeyStoreConfig.java
@@ -0,0 +1,30 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.common.settings;
+
+/**
+ * Configuration for the Elasticsearch keystore password.
+ * Returns the password from the KEYSTORE_PASSWORD environment variable,
+ * falling back to a BCFIPS-compliant default.
+ */
+public final class KeyStoreConfig {
+
+    private static final String DEFAULT_PASSWORD = "changeitchangeit";
+
+    private KeyStoreConfig() {}
+
+    public static char[] getPasswordChars() {
+        String password = System.getenv("KEYSTORE_PASSWORD");
+        if (password == null || password.isBlank()) {
+            password = DEFAULT_PASSWORD;
+        }
+        return password.toCharArray();
+    }
+}
diff --git a/server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java b/server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java
index a192dda84c4..03fb13f3c7f 100644
--- a/server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java
+++ b/server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java
@@ -221,13 +221,13 @@ public class KeyStoreWrapper implements SecureSettings {
         if (keystore != null && keystore.hasPassword()) {
             password = passwordSupplier.get();
         } else {
-            password = new SecureString(new char[0]);
+            password = new SecureString(KeyStoreConfig.getPasswordChars());
         }
 
         try (password) {
             if (keystore == null) {
                 final KeyStoreWrapper keyStoreWrapper = KeyStoreWrapper.create();
-                keyStoreWrapper.save(configDir, new char[0]);
+                keyStoreWrapper.save(configDir, KeyStoreConfig.getPasswordChars());
                 return keyStoreWrapper;
             } else {
                 keystore.decrypt(password.getChars());
-- 
2.52.0

